crates/driver/src/domain/competition/mod.rs

```diff
@@ -84,47 +86,102 @@ impl Competition {
+        settlements.shuffle(&mut rand::thread_rng());
+        // The merged settlements in their final form.
+        let mut results = Vec::new();
+        while let Some(settlement) = settlements.pop() {
+            // Has [`settlement`] been merged into another settlement?
+            let mut merged = false;
+            // Try to merge [`settlement`] into some other settlement.
+            for other in settlements.iter_mut() {
+                match other.merge(&settlement, &self.eth, &self.simulator).await {
+                    Ok(m) => {
+                        tracing::debug!(
+                            settlement_1 = ?settlement.solutions(),
+                            settlement_2 = ?other.solutions(),
+                            "merged solutions"
+                        );
+                        *other = m;
+                        merged = true;
+                        break;
+                    }
+                    Err(err) => {
+                        tracing::debug!(
+                            ?err,
+                            settlement_1 = ?settlement.solutions(),
+                            settlement_2 = ?other.solutions(),
+                            "solutions can't be merged"
+                        );
+                    }
+                }
+            }
+            // If [`settlement`] can't be merged into any other settlement, this is its
+            // final, most optimal form. Push it to the results.
+            if !merged {
+                results.push(settlement);
+            }
+        }
+
+        let settlements = results;
+
+        // Score the settlements.
+        let scores = settlements.into_iter().map(|settlement| {
+            tracing::trace!(
+                solutions = ?settlement.solutions(),
+                settlement_id = ?settlement.id,
+                "scoring settlement"
+            );
+            (settlement.score(&self.eth, auction), settlement)
+        });
+
+        // Filter out settlements which failed scoring.
+        let scores = scores.into_iter().filter_map(|(result, settlement)| {
+            result
+                .tap_err(|err| {
+                    tracing::info!(
+                        ?err,
+                        solutions = ?settlement.solutions(),
+                        "discarding settlement: failed scoring"
+                    );
+                })
+                .ok()
+                .map(|score| (score, settlement))
+        });
 
         // Trace the scores.
         let scores = scores.map(|(score, settlement)| {
-            tracing::info!(id = ?settlement.solution.id, score = f64::from(score.clone()), "solution scored");
+            tracing::info!(
+                solutions = ?settlement.solutions(),
+                settlement_id = ?settlement.id,
+                score = f64::from(score.clone()),
+                "settlement scored"
+            );
             (score, settlement)
         });
```

Comment:
Not a huge fan of this because `map` is lazy. Changes to the code 👇 can have unintended consequences where the tracing doesn't happen anymore. (i.e. its relying on the fact that `max_by_key` will collect the iterator which isn't exactly clear from a "code intent" point of view IMO).

Reply:
https://github.com/cowprotocol/services/pull/1490/commits/bd3e26de4e7594553c0bb8ed7e68b8fbfd21342d